lib/EnzymeCore/src/rules.jl

```diff
@@ -82,23 +82,79 @@ function _annotate_tt(@nospecialize(TT0))
     return ft, tt
 end
 
-function has_frule_from_sig(@nospecialize(TT); world=Base.get_world_counter())
+const MaybeMethodTableView = @static if isdefined(Core.Compiler, :method_table)
+    Union{Nothing,Core.Compiler.MethodTableView}
+else
+    Nothing
+end
+
+function has_frule_from_sig(@nospecialize(TT);
+                            world::UInt=Base.get_world_counter(),
+                            method_table::MaybeMethodTableView=nothing,
+                            caller::Union{Nothing,Core.MethodInstance}=nothing)
     ft, tt = _annotate_tt(TT)
     TT = Tuple{<:Annotation{ft}, Type{<:Annotation}, tt...}
-    isapplicable(forward, TT; world)
+    return isapplicable(forward, TT; world, method_table, caller)
 end
 
-function has_rrule_from_sig(@nospecialize(TT); world=Base.get_world_counter())
+function has_rrule_from_sig(@nospecialize(TT);
+                            world::UInt=Base.get_world_counter(),
+                            method_table::MaybeMethodTableView=nothing,
+                            caller::Union{Nothing,Core.MethodInstance}=nothing)
     ft, tt = _annotate_tt(TT)
     TT = Tuple{<:Config, <:Annotation{ft}, Type{<:Annotation}, tt...}
-    isapplicable(augmented_primal, TT; world)
+    return isapplicable(augmented_primal, TT; world, method_table, caller)
 end
 
 # Base.hasmethod is a precise match we want the broader query.
-function isapplicable(@nospecialize(f), @nospecialize(TT); world=Base.get_world_counter())
+function isapplicable(@nospecialize(f), @nospecialize(TT);
+                      world::UInt=Base.get_world_counter(),
+                      method_table::MaybeMethodTableView=nothing,
+                      caller::Union{Nothing,Core.MethodInstance}=nothing)
     tt = Base.to_tuple_type(TT)
     sig = Base.signature_type(f, tt)
-    return !isempty(Base._methods_by_ftype(sig, -1, world)) # TODO cheaper way of querying?
+    mt = ccall(:jl_method_table_for, Any, (Any,), sig)
+    mt isa Core.MethodTable || return false
+    @static if isdefined(Core.Compiler, :method_table)
```

Comment:
Keeping compatibility with Julia<1.8 is not very fun... Can Enzyme drop 1.6/1.7 compatibility?

Reply:
1.6 won't support custom rules. But folks have asked that we keep supporting 1.6 until the next LTS

Reply:
Yeah so we need rules on 1.7, 1.6 support here is fine to drop [since it won't have rules anyways]